Expose SAM tags for molecular indexes in IlluminaBasecallsToSam.

nh13 opened the pull request:
Add options to IlluminaBasecallsToSam to set the SAM tag for the
molecular index bases and base qualities.  Also added an option to allow
each molecular index to be stored in its own tag.

nh13 commented:
@yfarjoun when you are back from vacation, could you take a look at this one.  Or @eitanbanks if he remains in the wild, would you mind taking a look?

yfarjoun commented on src/java/picard/illumina/ClusterDataToSamConverter.java:
```diff
@@ -65,6 +67,10 @@
     private final int outputRecordsPerCluster;
     private final ReadNameEncoder readNameEncoder;
 
+    private String molecularIndexTag          = "RX";
```
perhaps add a todo to change this to reference SAMTagsUtil when I finally get around to changing the samspec: https://github.com/samtools/hts-specs/pull/119

nh13 replied:
done

yfarjoun commented on src/java/picard/illumina/ClusterDataToSamConverter.java:
```diff
@@ -134,12 +169,29 @@ private SAMRecord createSamRecord(final ReadData readData, final String readName
 
-        if (molecularIndex != null) {
+        if (!molecularIndexes.isEmpty()) {
             //TODO: add RX and QX to the list of SAMTags and change this. initial discussion
```
the todos here become very out of place..coudl you move them to where I commented above instead?

nh13 replied:
done

yfarjoun commented on src/java/picard/illumina/ClusterDataToSamConverter.java:
```diff
@@ -101,10 +107,39 @@ public ClusterDataToSamConverter(final String runBarcode,
+     *  Sets the SAM tag to use to store the molecular index bases.  If multiple molecular indexes exist, it will concatenate them
+     * and store them in this tag.
+     */
+    public ClusterDataToSamConverter withMolecularIndexTag(final String molecularIndexTag) {
```
I'm intregued by the use of the word "with" in the name of this method, when the verb is clearly "set"...why not the set* pattern we have in many other factories?

nh13 replied:
I like "with*" since it reads more naturally:

```
new ClusterDataToSamConverter(...).withMolecularIndexTag(...).withMolecularIndexQualityTag(...).withTagPerMolecularIndex(...)
```

versus

```
new ClusterDataToSamConverter(...).setMolecularIndexTag(...).setMolecularIndexQualityTag(...).setTagPerMolecularIndex(...)
```

Also, I don't see a lot of factories in Picard, three to be exact, so perhaps `with*` is reasonable?  As this is a style question, perhaps others @ktibbett or @tfenne could weigh in?

yfarjoun commented on src/java/picard/illumina/ClusterDataToSamConverter.java:
```diff
@@ -101,10 +107,39 @@ public ClusterDataToSamConverter(final String runBarcode,
+     * in the last tag.  If more tags are provided than molecular indexes found, the additional tags will not be used.
+     */
+    public ClusterDataToSamConverter withTagPerMolecularIndex(final List<String> tagPerMolecularIndex) {
+        if (tagPerMolecularIndex != null) this.tagPerMolecularIndex = tagPerMolecularIndex;
```
Should this throw a IllegalArgumentException rather than silently ignore a null being passed in?

nh13 replied:
Happy either way, I will do you ask.

yfarjoun replied:
I think that a null is an Illegal argument..so I'd vote for an exception.

yfarjoun commented on src/java/picard/illumina/ClusterDataToSamConverter.java:
```diff
@@ -134,12 +169,29 @@ private SAMRecord createSamRecord(final ReadData readData, final String readName
+            sam.setAttribute(this.molecularIndexQualityTag, String.join("", molecularIndexQualities));
+            if (!this.tagPerMolecularIndex.isEmpty() && molecularIndexes.size() > 1) {
+                if (tagPerMolecularIndex.size() < molecularIndexes.size()) {
+                    // fewer tags than indexes, so store the remaining in the last tag
```
This is a strange outcome that should result in a warning or an exception...

nh13 replied:
This is explained both in the definition of the `withTagPerMolecularIndex` method and in the `IlluminaBasecallsToSam`'s `TAG_PER_MOLECULAR_INDEX` option. 

What I *really* want is a read structure that can describe which molecular indexes should go into which tags (with concatenation), and possible ordering in concatenation.  For example `8M{RX}50T50T4M{ZA}4M{RX}` but that is for a different day. 

 Nonetheless, the convention of concatenating the molecular indexes when fewer tags are available is up for debate here.  I am in favor of this behavior.  Do others have thoughts?  @ktibbett @tfenne?

I will at minimum put a warning in place.

nh13 commented:
@yfarjoun thanks for the comments.  I'd like to get feedback from @ktibbett or @tfenne about two comments you made:
1. `with*` versus `set*`.
2. concatenation of the remaining molecular indexes when fewer tags than molecular indexes are given. 

I will push updated the branch once those two issues are resolved so we don't lose the comments :)

yfarjoun commented:
FYI: I'm not entirely done with the review, yet...just took a break for
dinner. :-)

On Sun, Apr 3, 2016 at 8:33 PM, Nils Homer <notifications@github.com> wrote:

> @yfarjoun <https://github.com/yfarjoun> thanks for the comments. I'd like
> to get feedback from @ktibbett <https://github.com/ktibbett> or @tfenne
> <https://github.com/tfenne> about two comments you made:
> 1. with* versus set*.
> 2. concatenation of the remaining molecular indexes when fewer tags than
> molecular indexes are given.
>
> I will push updated the branch once those two issues are resolved so we
> don't lose the comments :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/picard/pull/495#issuecomment-205085018>
>

eitanbanks commented:
(FYI I'm back online too and can help review if needed.  But I don't know nearly as much about molecular barcodes as Yossi...)

yfarjoun commented on src/java/picard/illumina/ClusterDataToSamConverter.java:
```diff
@@ -163,33 +215,29 @@ private SAMRecord createSamRecord(final ReadData readData, final String readName
-            final StringBuilder joinedMolecularIndexQBuilder = new StringBuilder();
-            final byte[][] molecularIndex = new byte[molecularBarcodeIndices.length][];
-            final byte[][] molecularIndexQ = new byte[molecularBarcodeIndices.length][];
+            molecularIndexes         = new ArrayList<>();
```
may I suggest better alignment here?

nh13 replied:
you may

yfarjoun commented on src/java/picard/illumina/ClusterDataToSamConverter.java:
```diff
@@ -163,33 +215,29 @@ private SAMRecord createSamRecord(final ReadData readData, final String readName
         } else {
-            joinedMolecularIndex = null;
-            joinedMolecularIndexQ = null;
+            molecularIndexes         = Collections.emptyList();
```
alignment

nh13 replied:
mapping </joke>

yfarjoun commented:
one over-all question: I'm wondering why you didn't do the same for the qualities...are you not interested in the qualities? if you're going to go through the trouble of parsing the QX field to get the qualities of each barcode, you could have done the same for the barcodes themselves...I'm not asking that you do this...just wondering why you didn't..

yfarjoun commented on src/java/picard/illumina/ClusterDataToSamConverter.java:
```diff
@@ -101,10 +107,39 @@ public ClusterDataToSamConverter(final String runBarcode,
+     * and store them in this tag.
+     */
+    public ClusterDataToSamConverter withMolecularIndexTag(final String molecularIndexTag) {
+        this.molecularIndexTag = molecularIndexTag;
```
what about null checking here?

this brings up a point: should we allow for a null `molecularIndexTag` indicating that you do not want the result stored concatenated in a tag but only in the  `tagPerMolecularIndex` ?

nh13 replied:
How about I check for a null or empty string?

yfarjoun replied:
Actually, now that you mention empty strings....

null -> error.
empty string -> no tag. 

would be my preference.

nh13 replied:
Oh how I wish java had `nonEmpty`.  Will do.

yfarjoun commented on src/java/picard/illumina/ClusterDataToSamConverter.java:
```diff
@@ -101,10 +107,39 @@ public ClusterDataToSamConverter(final String runBarcode,
+     * Sets the SAM tag to use to store the molecular index base qualities.  If multiple molecular indexes exist, it will concatenate them
+     * and store them in this tag.
+     */
+    public ClusterDataToSamConverter withMolecularIndexQualityTag(final String molecularIndexQualityTag) {
```
null check.

yfarjoun commented on src/java/picard/illumina/ClusterDataToSamConverter.java:
```diff
@@ -134,12 +169,29 @@ private SAMRecord createSamRecord(final ReadData readData, final String readName
-            sam.setAttribute("QX", molecularIndexQ);
+            sam.setAttribute(this.molecularIndexTag, String.join("", molecularIndexes));
+            sam.setAttribute(this.molecularIndexQualityTag, String.join("", molecularIndexQualities));
+            if (!this.tagPerMolecularIndex.isEmpty() && molecularIndexes.size() > 1) {
```
why the protection against `molecularIndexes.size() == 1` ?

nh13 replied:
I suppose I don't really need it, but the thinking was this would `act` the same as `molecularIndexTag`.  I'll remove it.

yfarjoun commented on src/java/picard/illumina/ClusterDataToSamConverter.java:
```diff
@@ -134,12 +169,29 @@ private SAMRecord createSamRecord(final ReadData readData, final String readName
+                }
+                else {
+                    // at least as many tags as indexes
+                    for (int i = 0; i < tagPerMolecularIndex.size(); i++) {
```
should be `i < molecularIndexes.size()`, no?

yfarjoun replied:
(this is what the request for the extra test should test)

nh13 replied:
gotcha

yfarjoun commented on src/tests/java/picard/illumina/IlluminaBasecallsToSamTest.java:
```diff
@@ -101,6 +101,27 @@ public void testNonBarcodedWithMoleclarIndex() throws Exception {
     }
 
+    // This *should not* store any molecular indexes in ZA or ZB since there is only one molecular index
+    @Test
```
please add a test with 3 tags and 2 barcodes

nh13 replied:
good idea, easy to do, done

yfarjoun commented on src/java/picard/illumina/IlluminaBasecallsToSam.java:
```diff
@@ -206,6 +206,16 @@
             "BAMs for only a subset of the barcodes in a lane.", shortName="INGORE_UNEXPECTED")
     public boolean IGNORE_UNEXPECTED_BARCODES = false;
 
+    @Option(doc="The tag to use to store any molecular indexes.  If more than one molecular index is found, they will be concatenated and stored here.")
```
optional?

nh13 replied:
Sure, so someone could give `null` to not populate the tag.

yfarjoun commented on src/java/picard/illumina/IlluminaBasecallsToSam.java:
```diff
@@ -206,6 +206,16 @@
+    @Option(doc="The tag to use to store any molecular indexes.  If more than one molecular index is found, they will be concatenated and stored here.")
+    public String MOLECULAR_INDEX_TAG = "RX";
+
+    @Option(doc="The tag to use to store any molecular index base qualities.  If more than one molecular index is found, their qualities will be concatenated and stored here.")
```
optional?

nh13 replied:
Same as above, thanks!

yfarjoun commented on src/java/picard/illumina/IlluminaBasecallsToSam.java:
```diff
@@ -206,6 +206,16 @@
+    @Option(doc="The tag to use to store any molecular index base qualities.  If more than one molecular index is found, their qualities will be concatenated and stored here.")
+    public String MOLECULAR_INDEX_BASE_QUALITY_TAG = "QX";
+
+    @Option(doc="The list of tags to store each molecular index.  If fewer tags are provided than molecular indexes, the remaining molecular indexes " +
```
optional?

nh13 replied:
Not needed since `minElements` defaults to zero.

nh13 replied:
But I will add optional and empty list since I am rusty with the clp parsing conventions in picard.

yfarjoun commented on src/java/picard/illumina/IlluminaBasecallsToSam.java:
```diff
@@ -206,6 +206,16 @@
+
+    @Option(doc="The list of tags to store each molecular index.  If fewer tags are provided than molecular indexes, the remaining molecular indexes " +
+            "will be concatenated and stored in the last tag.  If more tags are provided than molecular indexes found, the remaining tags will be ignored.")
+    public List<String> TAG_PER_MOLECULAR_INDEX;
```
default empty list?

nh13 replied:
will do, I am not sure if it is necessary.

yfarjoun commented on src/tests/java/picard/illumina/IlluminaBasecallsToSamTest.java:
```diff
@@ -117,7 +138,52 @@ public void testNonBarcodedWithDualMoleclarIndex() throws Exception {
         IOUtil.assertFilesEqual(outputBam, new File(TEST_DATA_DIR, "nonBarcodedWithMolecularIndex4M4M.sam"));
+    }
+
+    // This *should* store any molecular indexes in ZA and ZB since there is more than one molecular indexes
```
remove "any" from comment.

yfarjoun replied:
is->are in comment

yfarjoun commented on src/tests/java/picard/illumina/IlluminaBasecallsToSamTest.java:
```diff
@@ -117,7 +138,52 @@ public void testNonBarcodedWithDualMoleclarIndex() throws Exception {
+        IOUtil.assertFilesEqual(outputBam, new File(TEST_DATA_DIR, "nonBarcodedWithTagPerMolecularIndex4M4M.sam"));
+    }
 
+    // This *should* store any molecular indexes in ZA, ZB, and ZC since there is more than one molecular indexes.  There are
```
any, is

nh13 commented:
@yfarjoun I have addressed your comments and squashed (sorry).  I think the last thing that needs broader agreement is if we should concatenated additional molecular indexes in the last provided tag, which you thought was odd, and I thought was creative.

yfarjoun commented:
The problem with the concatenation is that is feels like a user error. I mean the user specified in the same invocation of the tool a different number of M operators in the ReadStructure than per-barcode-tags. do you see a use for this, or is it just a way around a possible mistake on the user's part? if the latter, I would prefer a inputValidation error upfront.

yfarjoun commented:
thinking about this more, I do not see any use-case in which the number of tags should be different than the number of M-operators. do you? If you agree, I'd say that the validation should enforce the same number, and the code will be simpler.

yfarjoun commented on src/java/picard/illumina/ClusterDataToSamConverter.java:
```diff
@@ -101,10 +115,40 @@ public ClusterDataToSamConverter(final String runBarcode,
+     */
+    public ClusterDataToSamConverter withTagPerMolecularIndex(final List<String> tagPerMolecularIndex) {
+        if (tagPerMolecularIndex == null) throw new IllegalArgumentException("Null given for tagPerMolecularIndex");
+        this.tagPerMolecularIndex = tagPerMolecularIndex;
```
according to sam-spec: "TAG is a two-character string that matches /[A-Za-z][A-Za-z0-9]/".. the code should probably enforce this in order to generate valid SAM. (x3)

nh13 replied:
`setAttribute` will do this for you :+1:

yfarjoun replied:
ok.

yfarjoun commented:
- with vs. set :+1:
- #tags vs. #M operators, unless you have a usecase (not a contrived one) where this is important, I'd advocate for a validation that enforces equality
- sam tags should be of the form set out in the sam-spec (please validate)

Thank-you for this PR.

nh13 commented:
@yfarjoun addressed all your comments.  I validated equal # of tags as molecular indices in both `IllluminaBasecallsToSam` and `ClusterDataToSamConverter`.  I didn't validate the sam tag spec, since that is handled elsewhere (in `setAttribute`).  If you :+1: I will rebase and squash.

yfarjoun commented:
thanks. :+1:  after minor comments.

nh13 commented:
@yfarjoun rebase and squashed.  Click the merge when tests pass?

yfarjoun commented:
Thanks!
